### PR TITLE
Add onValidate to Form

### LINKED ISSUE #3921: Access Validation Status of Form
Is there a way to access form validation state? If not, something like an `onValidate` event would be nice.

### PR DESCRIPTION
**Pull request by taysea**
#### What does this PR do?

This adds onValidate to form which allows users to access `error` and `info` when a form validates. This would allow users to access which formfields have errors prior to form submission. 

Additionally, `errors` and `infos` had been stored separately, but are now combined into a single `validationResults` state object that contains them both since validation runs on both.

This helps address issues noted in these use cases:
https://github.com/hpe-design/design-system/pull/998
https://github.com/grommet/grommet/issues/4375

#### Where should the reviewer start?
src/js/components/Form/Form.js

#### What testing has been done on this PR?
Added tests in jest

Added example in `AllChildren` Form storybook.

#### How should this be manually tested?

add this to a form:
`onValidate={({error, info}) => console.log(error, info)}`

#### Any background context you want to provide?
Users didn't have access to the error and info data that Form holds in state.

#### What are the relevant issues?
Closes https://github.com/hpe-design/design-system/issues/1047
Closes #3921 
#### Do the grommet docs need to be updated?
Yes, updated in this PR.

#### Should this PR be mentioned in the release notes?
Yes.

#### Is this change backwards compatible or is it a breaking change?
Yes, backwards compatible.

### CODE REVIEW

File: src/js/components/Form/Form.js [hunk 1]
```diff
@@ -233,6 +234,14 @@ const Form = forwardRef(
                 // which probably came from a submit
                 setErrors(prevErrors => ({ ...prevErrors, ...nextErrors }));
                 setInfos(prevInfos => ({ ...prevInfos, ...nextInfos }));
+
+                // give user access to errors that have occurred on validation
+                if (onValidate) {
+                  const validation = {};
+                  validation.error = { ...errors, ...nextErrors };
```
Comment (ericsoderberghp) [inline src/js/components/Form/Form.js:241]: It feels slightly odd that we are using `errors` here but `prevErrors` above on line 235. I wonder if we should merge `errors` and `infos` into the same state. For example:
```
const [validationResults, setValidationResults] = useState(defaultValidationResults);
...
setValidationResults(prevValidationResults => {
  const nextValidationResults = { errors: { ...prevValidationResults.errors, ...nextErrors }, infos: ... };
  if (onValidate) onValidate(nextValidationResults);
  return nextValidationResults;
});
```

File: src/js/components/Form/Form.js [hunk 2]
```diff
@@ -233,6 +234,14 @@ const Form = forwardRef(
...
+                  validation.info = { ...infos, ...nextInfos };
+                  onValidate(validation);
```
Comment (ericsoderberghp) [inline src/js/components/Form/Form.js:243]: It seems a bit inconsistent to use the `event` for `onSubmit` but not for `onBlur`.

File: src/js/components/Form/doc.js [hunk 3]
```diff
@@ -45,6 +45,11 @@ export const doc = Form => {
       `Function that will be called when the form is reset. The
       single argument is the event provided by react.`,
     ),
+    onValidate: PropTypes.func.description(
+      `Function that will be called when the form is validated. The
+      single argument is an event containing the latest error object
+      via \`event.error\` and info object via \`event.info\`.`,
```
Comment (ericsoderberghp) [inline src/js/components/Form/doc.js:51]: I think these should be `event.errors` and `event.infos` to align with the property names.

Comment (taysea) [reply src/js/components/Form/doc.js:51]: I think that's a good idea

File: src/js/components/Form/Form.js (hunk 2 above)
Comment (taysea) [reply src/js/components/Form/Form.js:243]: I would agree. Updating `onBlur` to use the event as well.

File: src/js/components/Form/Form.js (hunk 1 above)
Comment (taysea) [reply src/js/components/Form/Form.js]: This is a nice idea. Let me try that

**Comment by ShimiSun** [comment]
Can you take a look at the Chormatic file change and confirm it is expected https://www.chromatic.com/snapshot?appId=5d9774839a6eff00203f5cbf&id=5f33215e912024002272fa83?

**Comment by taysea** [comment]
> Can you take a look at the Chormatic file change and confirm it is expected https://www.chromatic.com/snapshot?appId=5d9774839a6eff00203f5cbf&id=5f33215e912024002272fa83?

Yes, this is expected. I had added one formfield into that story to test an error message, so the visual change is due to a field being added to the form.

**Comment by ShimiSun** [comment]
Were you able to test this solution in the long-form context on the design-system? I'm curious whether it answered all the expected requirements.

**Comment by taysea** [comment]
> Were you able to test this solution in the long-form context on the design-system? I'm curious whether it answered all the expected requirements.

I have not implemented it, but I can try it out in this branch locally. @britt6612 had two good proposed approaches, so let me test them out here.

File: src/js/components/Form/Form.js [hunk 4]
```diff
@@ -56,21 +59,29 @@ const Form = forwardRef(
...
+      () => setValidationResults({ errors: errorsProp, infos: infosProp }),
+      [errorsProp, infosProp],
+    );
     const validations = useRef({});
 
     // clear any errors when value changes
     useEffect(() => {
-      setErrors(prevErrors => {
+      setValidationResults(prevValidationResults => {
         const [nextErrors] = validate(
```
Comment (ericsoderberghp) [inline src/js/components/Form/Form.js:75]: Should this be `[nextErrors, nextInfos]` and then use `nextInfos` at line 83? I would think the `validate()` could return some fresh infos.

Comment (taysea) [reply src/js/components/Form/Form.js:75]: I think you're right -- we didn't have that prior but I would also think that there could be some new infos coming in after validation

**Pull request merged**